### ISSUE TITLE
Canon TR4500 series rejects any request...

### ISSUE THREAD

**Issue by booti386** [OpenPrinting/ipp-usb#29]
... that has not the Host header set to either localhost:[port] or 127.0.0.1:[port]

Examples of success:
```
10-05-2021 17:14:47: > HTTP[005]: POST http://127.0.0.1:60001/ipp/print
10-05-2021 17:14:47: > HTTP[005]: request body: got 170 bytes; EOF
10-05-2021 17:14:47: > HTTP[005]: body is small (170 bytes), prefetched before sending
10-05-2021 17:14:47: > HTTP[005]: HTTP request header:
10-05-2021 17:14:47: >   POST /ipp/print HTTP/1.1
10-05-2021 17:14:47: >   Host: 127.0.0.1:60001
10-05-2021 17:14:47: >   User-Agent: CUPS/2.3.3 (Linux 5.8.0-52-generic; x86_64) IPP/2.0
10-05-2021 17:14:47: >   Content-Length: 170
10-05-2021 17:14:47: >   Accept-Encoding: deflate, gzip, identity
10-05-2021 17:14:47: >   Content-Type: application/ipp
10-05-2021 17:14:47: >   Date: Mon, 10 May 2021 15:14:47 GMT
10-05-2021 17:14:47: >
10-05-2021 17:14:47:   USB[1]: connection allocated, 1 in use: --- a--
10-05-2021 17:14:47:   HTTP[005]: connection 1 allocated
10-05-2021 17:14:47: > USB[1]: write: wanted 417 sent 417 total 417
10-05-2021 17:14:47: < USB[1]: read: wanted 4096 got 17 total 17
10-05-2021 17:14:47: < USB[1]: read: wanted 4096 got 19 total 36
10-05-2021 17:14:47: < USB[1]: read: wanted 4096 got 28 total 64
10-05-2021 17:14:47: < USB[1]: read: wanted 4096 got 81 total 145
10-05-2021 17:14:47: < HTTP[005]: POST http://127.0.0.1:60001/ipp/print - 200 OK
10-05-2021 17:14:47: < HTTP[005]: HTTP response header:
10-05-2021 17:14:47: <   HTTP/1.1 200 OK
10-05-2021 17:14:47: <   Connection: Keep-Alive
10-05-2021 17:14:47: <   Content-Type: application/ipp
10-05-2021 17:14:47: <   Keep-Alive: timeout=30
10-05-2021 17:14:47: <   Mime-Version: 1.0
10-05-2021 17:14:47: <   Transfer-Encoding: chunked
10-05-2021 17:14:47: <
10-05-2021 17:14:47: < USB[1]: read: wanted 4096 got 5 total 150
```
```
10-05-2021 17:17:31: > HTTP[003]: POST http://localhost:60001/ipp/print
10-05-2021 17:17:31: > HTTP[003]: request body: got 170 bytes; EOF
10-05-2021 17:17:31: > HTTP[003]: body is small (170 bytes), prefetched before sending
10-05-2021 17:17:31: > HTTP[003]: HTTP request header:
10-05-2021 17:17:31: >   POST /ipp/print HTTP/1.1
10-05-2021 17:17:31: >   Host: localhost:60001
10-05-2021 17:17:31: >   User-Agent: CUPS/2.3.3 (Linux 5.8.0-52-generic; x86_64) IPP/2.0
10-05-2021 17:17:31: >   Content-Length: 170
10-05-2021 17:17:31: >   Accept-Encoding: deflate, gzip, identity
10-05-2021 17:17:31: >   Content-Type: application/ipp
10-05-2021 17:17:31: >   Date: Mon, 10 May 2021 15:17:31 GMT
10-05-2021 17:17:31: >
10-05-2021 17:17:31:   USB[1]: connection allocated, 1 in use: --- a--
10-05-2021 17:17:31:   HTTP[003]: connection 1 allocated
10-05-2021 17:17:31: > USB[1]: write: wanted 417 sent 417 total 417
10-05-2021 17:17:31: < USB[1]: read: wanted 4096 got 17 total 17
10-05-2021 17:17:31: < USB[1]: read: wanted 4096 got 19 total 36
10-05-2021 17:17:31: < USB[1]: read: wanted 4096 got 28 total 64
10-05-2021 17:17:31: < USB[1]: read: wanted 4096 got 81 total 145
10-05-2021 17:17:31: < HTTP[003]: POST http://localhost:60001/ipp/print - 200 OK
10-05-2021 17:17:31: < HTTP[003]: HTTP response header:
10-05-2021 17:17:31: <   HTTP/1.1 200 OK
10-05-2021 17:17:31: <   Connection: Keep-Alive
10-05-2021 17:17:31: <   Content-Type: application/ipp
10-05-2021 17:17:31: <   Keep-Alive: timeout=30
10-05-2021 17:17:31: <   Mime-Version: 1.0
10-05-2021 17:17:31: <   Transfer-Encoding: chunked
10-05-2021 17:17:31: <
10-05-2021 17:17:31: < USB[1]: read: wanted 4096 got 5 total 150
[...]
```

Examples of failure:
```
10-05-2021 17:18:41: > HTTP[004]: POST http://127.0.0.2:60001/ipp/print
10-05-2021 17:18:41: > HTTP[004]: request body: got 170 bytes; EOF
10-05-2021 17:18:41: > HTTP[004]: body is small (170 bytes), prefetched before sending
10-05-2021 17:18:41: > HTTP[004]: HTTP request header:
10-05-2021 17:18:41: >   POST /ipp/print HTTP/1.1
10-05-2021 17:18:41: >   Host: 127.0.0.2:60001
10-05-2021 17:18:41: >   User-Agent: CUPS/2.3.3 (Linux 5.8.0-52-generic; x86_64) IPP/2.0
10-05-2021 17:18:41: >   Content-Length: 170
10-05-2021 17:18:41: >   Accept-Encoding: deflate, gzip, identity
10-05-2021 17:18:41: >   Content-Type: application/ipp
10-05-2021 17:18:41: >   Date: Mon, 10 May 2021 15:18:41 GMT
10-05-2021 17:18:41: >
10-05-2021 17:18:41:   USB[0]: connection allocated, 1 in use: a-- ---
10-05-2021 17:18:41:   HTTP[004]: connection 0 allocated
10-05-2021 17:18:41: > USB[0]: write: wanted 417 sent 417 total 417
10-05-2021 17:18:41: < USB[0]: read: wanted 4096 got 26 total 26
10-05-2021 17:18:41: < USB[0]: read: wanted 4096 got 19 total 45
10-05-2021 17:18:41: < USB[0]: read: wanted 4096 got 28 total 73
10-05-2021 17:18:41: < USB[0]: read: wanted 4096 got 19 total 92
10-05-2021 17:18:41: < USB[0]: read: wanted 4096 got 2 total 94
10-05-2021 17:18:41: < HTTP[004]: POST http://127.0.0.2:60001/ipp/print - 400 Bad Request
10-05-2021 17:18:41: < HTTP[004]: HTTP response header:
10-05-2021 17:18:41: <   HTTP/1.1 400 Bad Request
10-05-2021 17:18:41: <   Mime-Version: 1.0
10-05-2021 17:18:41: <   Transfer-Encoding: chunked
10-05-2021 17:18:41: <
10-05-2021 17:18:41: < USB[0]: read: wanted 4096 got 5 total 99
10-05-2021 17:18:41: < HTTP[004]: response body: got 0 bytes; EOF
```
```
10-05-2021 17:19:24: > HTTP[005]: POST http://localhost.localdomain:60001/ipp/print
10-05-2021 17:19:24: > HTTP[005]: request body: got 182 bytes; EOF
10-05-2021 17:19:24: > HTTP[005]: body is small (182 bytes), prefetched before sending
10-05-2021 17:19:24: > HTTP[005]: HTTP request header:
10-05-2021 17:19:24: >   POST /ipp/print HTTP/1.1
10-05-2021 17:19:24: >   Host: localhost.localdomain:60001
10-05-2021 17:19:24: >   User-Agent: CUPS/2.3.3 (Linux 5.8.0-52-generic; x86_64) IPP/2.0
10-05-2021 17:19:24: >   Content-Length: 182
10-05-2021 17:19:24: >   Accept-Encoding: deflate, gzip, identity
10-05-2021 17:19:24: >   Content-Type: application/ipp
10-05-2021 17:19:24: >   Date: Mon, 10 May 2021 15:19:24 GMT
10-05-2021 17:19:24: >
10-05-2021 17:19:24:   USB[1]: connection allocated, 1 in use: --- a--
10-05-2021 17:19:24:   HTTP[005]: connection 1 allocated
10-05-2021 17:19:24: > USB[1]: write: wanted 441 sent 441 total 441
10-05-2021 17:19:24: < USB[1]: read: wanted 4096 got 26 total 26
10-05-2021 17:19:24: < USB[1]: read: wanted 4096 got 19 total 45
10-05-2021 17:19:24: < USB[1]: read: wanted 4096 got 28 total 73
10-05-2021 17:19:24: < USB[1]: read: wanted 4096 got 19 total 92
10-05-2021 17:19:24: < USB[1]: read: wanted 4096 got 2 total 94
10-05-2021 17:19:24: < HTTP[005]: POST http://localhost.localdomain:60001/ipp/print - 400 Bad Request
10-05-2021 17:19:24: < HTTP[005]: HTTP response header:
10-05-2021 17:19:24: <   HTTP/1.1 400 Bad Request
10-05-2021 17:19:24: <   Mime-Version: 1.0
10-05-2021 17:19:24: <   Transfer-Encoding: chunked
10-05-2021 17:19:24: <
10-05-2021 17:19:24: < USB[1]: read: wanted 4096 got 5 total 99
10-05-2021 17:19:24: < HTTP[005]: response body: got 0 bytes; EOF
```

For information:
```
$ host localhost
localhost has address 127.0.0.1
$ host localhost.localdomain
localhost.localdomain has address 127.0.0.1
$ 
```

**Comment by alexpevzner** [comment]
Cc: @tillkamppeter, @michaelrsweet

Unfortunately, this problem cannot be fixed here. Under some circumstances Avahi announces localhost devices under name different from `localhost`. `ipp-usb` cannot control it.

Most of the printers still work in this situation, but few models are more restrictive.

Ideally, it should be fixed at the Avahi side. But experience shows that getting something fixed at Avahi is nearly hopeless. Alternatively, workaround could be added to CUPS to always set `Host: localhost:port`, if printer's address is `127.0.0.1` or `::1`, regardless of printer's host name, obtained from Avahi

**Comment by booti386** [comment]
I tried to fix it by creating a quirk for ipp-usb (`http-host = localhost:6000`), but it didn't work and the header value remained the same in the logs.
I also tried with `http-text = bla` and this time the test header properly showed up in the logs.

I suspect it's because the http backend overwrites automagically the host field by default, but couldn't it be useful to add the ability to override it anyway using quirks?

**Comment by alexpevzner** [comment]
@booti386,

this issue cannot be fixed as in #30. The problem is that responses to some IPP requests may contains URLs, and because in the IPP over USB case device has no idea regardless its hostname, it simply mirrors hostname from the `Host:` header of request.

If we override the `Host:` reader in request, URLs response will point to different domain that request was sent to. It may cause unobvious problems.
